### DISCUSSION TITLE
PIP-27: Pruned Node

### DISCUSSION THREAD

**Discussion by ja7ad** [pactus-project/PIPs#134] [PIPs]
PIP-27: [Pruned Node](https://pips.pactus.org/PIPs/pip-27)

**Comment by kehiy** [comment]
@Ja7ad @b00f @pactus-project/research-reviewer  I think it's better to call it light node or light client since every protocol did it. I mean everyone knows this concept as the light client so it can cause misunderstanding.

**Reply by b00f** [reply]
In my opinion, a light client should refer to a type of client that doesn't verify everything—maybe only new blocks without validating transactions.

**Reply by kehiy** [reply]
I agree with that.

**Comment by kehiy** [comment]
@Ja7ad @b00f @pactus-project/research-reviewer 

https://github.com/pactus-project/PIPs/blob/35f87692233fb0744188c58ac13caf6e3e8b7bfe/PIPs/pip-27.md?plain=1#L87

In this case, the pruned node is downloading only some part of the chain! not the whole chain from Genesis, so can we still trust the centralized server by verifying?

We are only verifying and seeing some parts of the chain, is there any way for manipulation by a centralized server?

And one more important issue, we are syncing some blocks at all, what about indexed data extracted from transactions? balances, stakes, and in future contracts states?

When we have a whole chain, we can retrieve states by reading whole blocks and indexing the changes that a transaction makes, such as bonding a new validator, sending a native coin, or a contract deploy.

But when we don't have the whole chain we need these data and we need to verify them as well.

**Comment by b00f** [comment]
If the data you receive from the centralized server is manipulated, you can't sync the data and the chain. 
The specification should be modified. Instead of copying the data, we have to iterate over blocks to ensure all the block have valid signature and in the last block the state hash is the same as sate hash inside the block.

**Reply by kehiy** [reply]
How it can calculate the state hash when the light node doesn't have the old blocks and states?

Such as balances or stake amounts? One way is to sync with the whole chain for 1 time and then remove old blocks.

But keeping the state without blocks can still be a lot if we are thinking about storage.

Also, I think the only thing reduced in this proposal is the storage and the process cost and network cost are the same.

With a simple compare and looking at blow chart for example, we can find out the storage is not a serious issue, at most this pruned node can make it half. (for now makes 500Mb -> 250Mb)

![photo_6026286665221391203_y](https://github.com/pactus-project/PIPs/assets/89645414/8f60e57c-2622-4d66-a607-88dca17c7b63)


![photo_6026286665221391204_x](https://github.com/pactus-project/PIPs/assets/89645414/7657b6a6-74f6-4736-b8aa-b58b7f71d0c4)

**Comment by b00f** [comment]
> How it can calculate the state hash when the light node doesn't have the old blocks and states?

Once you have the latest state of blockchain, including all accounts and validatos, you can calculate the state hash. [Here](https://docs.pactus.org/protocol/blockchain/state-hash/)

> But keeping the state without blocks can still be a lot if we are thinking about storage.

Agree, but still we have less data to store.

**Reply by kehiy** [reply]
> Once you have the latest state of blockchain, including all accounts and validatos, you can calculate the state hash. [Here](https://docs.pactus.org/protocol/blockchain/state-hash/)

So, we need to sync with the state too, correct? In this case, this is a good model.

About storage, I think I was wrong. we have much less data. we don't have all TXs changing 1 balance integer. we keep only an integer for a balance and it means less data.

But one more question here: A pruned node can't really prove the state it has based on history. like asking where this 1000 PAC comes into this smart contract or validator? so it can't provide a detailed history of it.

Of course it's not required for pruned nodes and we don't expect it, but can it cause security issues?

**Comment by ja7ad** [comment]
PIP-27 for implementation accepted, Issue address is: https://github.com/pactus-project/pactus/issues/1390

**Comment by b00f** [comment]
@kehiy 

Let's remove the check on downloaded data that seems redundant to me. 
Consider an adversary who can change all blocks and transactions up to the last one. 
After 10 days, the corrupted block will be pruned, giving no advantage to the adversary.

**Reply by b00f** [reply]
Please review: https://github.com/pactus-project/PIPs/pull/164

**Reply by kehiy** [reply]
@b00f That's OK, but we must keep one thing on PIP.

We need to keep `verify` or your explanation here about why we don't need to verify if we don't say anything about it it's not logical.

**Reply by kehiy** [reply]
I Checked the changes on PIP now, looks good to me.
